src/super_gradients/training/losses/structure_loss.py

```diff
@@ -0,0 +1,118 @@
+
+import torch
+from torch.nn.modules.loss import _Loss
+
+from super_gradients.common.abstractions.abstract_logger import get_logger
+from super_gradients.training.losses.loss_utils import apply_reduce, LossReduction
+from super_gradients.training.utils.segmentation_utils import to_one_hot
+
+logger = get_logger(__name__)
+
+
+class AbstarctStructureLoss(_Loss, ABC):
+    """
+    Abstract computation of structure loss between two tensors, It can support both multi-classes and binary tasks.
+    """
+    def __init__(self,
+                 apply_softmax: bool = True,
+                 ignore_index: int = None,
+                 smooth: float = 1.,
+                 eps: float = 1e-5,
+                 sum_over_batches: bool = False,
+                 generalized_metric: bool = False,
+                 weight: Optional[torch.Tensor] = None,
+                 reduction: Union[LossReduction, str] = "mean"):
+        """
+        :param apply_softmax: Whether to apply softmax to the predictions.
+        :param smooth: laplace smoothing, also known as additive smoothing. The larger smooth value is, closer the metric
+            coefficient is to 1, which can be used as a regularization effect.
+            As mentioned in: https://github.com/pytorch/pytorch/issues/1249#issuecomment-337999895
+        :param eps: epsilon value to avoid inf.
+        :param sum_over_batches: Whether to average metric over the batch axis if set True,
+         default is `False` to average over the classes axis.
+        :param generalized_metric: Whether to apply normalization by the volume of each class.
+        :param weight: a manual rescaling weight given to each class. If given, it has to be a Tensor of size `C`.
+        :param reduction: Specifies the reduction to apply to the output: `none` | `mean` | `sum`.
+            `none`: no reduction will be applied.
+            `mean`: the sum of the output will be divided by the number of elements in the output.
+            `sum`: the output will be summed.
+            Default: `mean`
+        """
+        super().__init__(reduction=reduction)
+        self.ignore_index = ignore_index
+        self.apply_softmax = apply_softmax
+        self.eps = eps
+        self.smooth = smooth
+        self.sum_over_batches = sum_over_batches
+        self.generalized_metric = generalized_metric
+        self.weight = weight
+        if self.generalized_metric:
+            assert self.weight is None, "Cannot use structured Loss with weight classes and generalized normalization"
+            if self.eps > 1e-12:
+                logger.warning("When using GeneralizedLoss, it is recommended to use eps below 1e-12, to not affect"
+                               "small values normalized terms.")
+            if self.smooth != 0:
+                logger.warning("When using GeneralizedLoss, it is recommended to set smooth value as 0.")
+
+    @abstractmethod
+    def _calc_numerator_denominator(self, labels_one_hot, predict):
```

Comment:
please add `-> (torch.Tensor, torch.Tensor)` to signature

Reply:
added b863160d123d7e618f80e4beab8b51472e3e64e4